src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Binders/HeaderModelBinder.cs

```diff
@@ -45,26 +50,112 @@ public HeaderModelBinder(ILoggerFactory loggerFactory)
+        }
+
+        // to enable unit testing
+        internal IModelBinder InnerModelBinder { get; }
+
         /// <inheritdoc />
-        public Task BindModelAsync(ModelBindingContext bindingContext)
+        public async Task BindModelAsync(ModelBindingContext bindingContext)
         {
             if (bindingContext == null)
             {
                 throw new ArgumentNullException(nameof(bindingContext));
             }
 
-            var request = bindingContext.HttpContext.Request;
+            _logger.AttemptingToBindModel(bindingContext);
 
             // Property name can be null if the model metadata represents a type (rather than a property or parameter).
             var headerName = bindingContext.FieldName;
 
-            _logger.AttemptingToBindModel(bindingContext);
-
+            var request = bindingContext.HttpContext.Request;
             if (!request.Headers.ContainsKey(headerName))
             {
                 _logger.FoundNoValueInRequest(bindingContext);
             }
 
+            if (InnerModelBinder == null)
+            {
+                BindWithoutInnerBinder(bindingContext);
+                return;
+            }
+
+            // Do not set ModelBindingResult to Failed on not finding the value in the header as we want the inner 
+            // modelbinder to do that. This would give a chance to the inner binder to add more useful information.
+            // For example, SimpleTypeModelBinder adds a model error when binding to let's say an integer and the
+            // model is null.
+
+            var headerValueProvider = GetHeaderValueProvider(headerName, bindingContext);
+
+            // Create a new binding scope in order to supply the HeaderValueProvider so that the binders like
+            // SimpleTypeModelBinder can find values from header.
+            ModelBindingResult result;
+            using (bindingContext.EnterNestedScope(
+                    bindingContext.ModelMetadata,
+                    fieldName: bindingContext.FieldName,
+                    modelName: bindingContext.ModelName,
+                    model: bindingContext.Model))
+            {
+                bindingContext.ValueProvider = headerValueProvider;
+
+                await InnerModelBinder.BindModelAsync(bindingContext);
+                result = bindingContext.Result;
+            }
+
+            bindingContext.Result = result;
+
+            _logger.DoneAttemptingToBindModel(bindingContext);
+        }
+
+        private HeaderValueProvider GetHeaderValueProvider(string headerName, ModelBindingContext bindingContext)
+        {
+            var request = bindingContext.HttpContext.Request;
+
+            // Prevent breaking existing users in scenarios where they are binding to a 'string' property
+            // and expect the whole comma separated string, if any, as a single string and not as a string array.
+            string[] values;
```

Comment:
May be a small perf win to avoid allocating this array when `!request.Headers.ContainsKey(headerName)`. Just use `Array.Empty<string>()` in that case.